[Speculative Decoding] Medusa Implementation with Top-1 proposer

abhigoyal1997 opened the pull request:
This PR implements the Medusa approach to generate speculations using top-1 predictions of the heads.

For Mistral-7B-Instruct-v0.2 and Meta-Llama-3-8B-Instruct on an H100 card, the following are the throughput numbers (total tokens generated/sec) when tested on MT-Bench:

  | Temperature |   |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | --
Concurrency |   | 64 | 32 | 16 | 8 | 4 | 2 | 1
Mistral-7B-Instruct-v0.2 | 0 | 2425.03 | 1862.19 | 1262.11 | 792.05 | 453.78 | 245.14 | 128.73
  | 1 | 1876.1 | 1902.74 | 1281.7 | 799.73 | 452.16 | 245.8 | 126.73
  |   |   |   |   |   |   |   |  
Mistral-7B-Instruct-v0.2 + medusa | 0 | 1769.62 | 1681.17 | 1318.1 | 911.77 | 550.85 | 293.14 | 155.06
  | 1 | 1818.85 | 1739.18 | 1414.85 | 911.56 | 554.67 | 297.65 | 164.38
  |   |   |   |   |   |   |   |  
Meta-Llama-3-8B-Instruct | 0 | 2275.64 | 1796.47 | 1258.82 | 760.38 | 429.29 | 235.47 | 121.27
  | 1 | 2291.75 | 1712.22 | 1246.62 | 740.42 | 420.49 | 236.39 | 121.05
  |   |   |   |   |   |   |   |  
Meta-Llama-3-8B-Instruct + medusa | 0 | 1533.34 | 1445.23 | 1156.08 | 830.04 | 490.86 | 269.09 | 141.33
  | 1 | 1606.42 | 1457.75 | 1185.55 | 838.49 | 512.31 | 281.95 | 147.08

So for smaller batch sizes, we see improvement in tokens generated per sec.
Medusa heads for these models were trained using a set of public instructions. I am working on making those available via Huggingface Hub as well.

With tree-style speculation and verification, this should give even higher improvements.

FIX #1023
FIX #4669 
Fix FasterDecoding/Medusa#41

caddfa31434 commented:
@abhigoyal1997  Awesome PR.Do you have any plans regarding support for tree-style speculation and verification?

abhigoyal1997 commented:
Hi @caddfa31434 
Tree-style speculation and verification are certainly something I am interested in. But for complete benefit, I was still waiting on the support for tree attention. I have implemented it in a PyTorch implementation but haven't thought about how it would work in vLLM.

zhyncs commented:
> Hi @caddfa31434 Tree-style speculation and verification are certainly something I am interested in. But for complete benefit, I was still waiting on the support for tree attention. I have implemented it in a PyTorch implementation but haven't thought about how it would work in vLLM.

Hi @abhigoyal1997 We recently had the experience of developing a TreeMask-based version of Medusa on our internal inference framework, and it is expected to be ready by the end of the month. Here are some suggestions.

There are three main differences between the TreeMask version and the non-TreeMask version (three modification points):
1. The Cos/Sin matrix for RoPE
2. Replacing Casual Mask with Tree Mask in Attention
3. Updating KV Cache, which needs to be **compacted** after each verification

Additionally, EAGLE is a technology worth considering for selection. Therefore, I suggest you make some abstractions when implementing Medusa, which will make future integration with EAGLE easier.

abhigoyal1997 commented:
@zhyncs Thanks for the insights into tree-mask-based version.
As for EAGLE, I have an implementation ready for that as well. This implementation of Medusa does keep required abstractions in mind such that any speculative models which need some inputs from the target model can be implemented with ease. In fact the implementation of EAGLE I have (I'll raise a PR for that as well soon), only needs addition of a new model to vLLM without any change to the implementation logic of speculative decoding itself.

abhigoyal1997 commented:
@cadedaniel This is complete. Can you please review it?

cadedaniel commented:
Thanks for the contribution! Will take a look today or tomorrow.

cadedaniel commented:
+1; I suggest we generalize top-1 and top-k proposing scoring (including defragmentation of accepted KV). then we can use top-1 and top-k implementations with different spec proposal methods (draft, medusa, eagle, ngram, etc).

also, we can wait for masking in kernels; we can also implement it in batch expansion style. it won't be as performant but could be a faster way to get everything built as we can add in the kernel support when it's ready.

cc @LiuXiaoxuanPKU

zhyncs commented:
> we can also implement it in batch expansion style

The performance will be particularly poor. I don't recommend doing this.

zhyncs commented:
> > we can also implement it in batch expansion style
> 
> The performance will be particularly poor. I don't recommend doing this.

At the same time, supporting TreeMask in the Attention Kernel is not as much work as imagined. The prerequisite is to understand the implementation of the original casual mask. If you're interested, we can discuss the details further.

zhyncs commented:
> Especially when large batch size is considered, different request may have different acceptance path in the candidate tree.

From our internal experience, the real challenge is not integrating with continuous batching, but rather compatibility with existing features such as chunked prefill.

zhyncs commented:
> This efficient tree attention kernel will be a crucial factor in tree-style speculative decoding, which we should consider prioritize.

In fact, our implementation is not based on [Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention), but on the [TurboMind 2.1 Attention Kernel](https://github.com/InternLM/lmdeploy/tree/main/src/turbomind/kernels/attention), which was written from scratch and its performance is about 10% better than [Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention).

The changes related to casual mask are roughly as follows:
```cuda
// original
    __device__ void ApplyCasualMask(FragS& frag_S, int offset_Q, int offset_K)
    {
        Impl::ForeachS(frag_S, [&](int hi, int qi, int si, int ri, float& score) {
            if (offset_Q + qi < offset_K + si) {
                score -= std::numeric_limits<float>::infinity();
            }
        });
    }
```

```cuda
// modified
    __device__ void ApplyCasualMask(
        FragS& frag_S, int offset_Q, int offset_K, const int* medusa_mask, int his_len, int input_len, int query_idx)
    {
        Impl::ForeachS(frag_S, [&](int hi, int qi, int si, int ri, float& score) {
            if (medusa_mask) {
                int rel_pos_q = offset_Q + qi - his_len;
                int rel_pos_k = offset_K + si - his_len;
                if (0 <= rel_pos_q && rel_pos_q < input_len && 0 <= rel_pos_k && rel_pos_k < input_len) {
                    if (medusa_mask[rel_pos_q * input_len + rel_pos_k] == 0) {
                        score -= std::numeric_limits<float>::infinity();
                    }
                }
                else {
                    if (offset_Q + qi < offset_K + si) {
                        score -= std::numeric_limits<float>::infinity();
                    }
                }
            }
            else {
                if (offset_Q + qi < offset_K + si) {
                    score -= std::numeric_limits<float>::infinity();
                }
            }
        });
    }
```

zhyncs commented:
> The incompatibility mentioned here is more of a trade-off

This is related to the implementation of different frameworks, such as whether the previous framework has a sufficiently good abstract design and whether it is convenient for future expansion. It cannot be generalized. In comparison, vLLM is indeed more user-friendly in terms of secondary development difficulty and also has relatively strong scalability.

cadedaniel commented:
If you're interested in combining chunked prefill and spec decode, see https://github.com/vllm-project/vllm/issues/5016. We have a naive dynamic speculation length policy which disables spec decode when the batch size gets too large.

abhigoyal1997 commented:
Hi @cadedaniel 
Did you get a chance to look at this?

zhyncs commented:
> also, we can wait for masking in kernels

@cadedaniel FlashInfer supports custom mask now https://github.com/flashinfer-ai/flashinfer/pull/266

zhyncs commented:
> Hi @cadedaniel Did you get a chance to look at this?

@abhigoyal1997 May you resolve the conflicts first. Thanks.

abhigoyal1997 commented:
@zhyncs Thanks for pointing out. I've resolved the conflicts 👍

abhigoyal1997 commented on vllm/transformers_utils/config.py:
```diff
@@ -15,8 +19,13 @@
+    "medusa": MedusaConfig,
 }
 
+for name, cls in _CONFIG_REGISTRY.items():
```
Without this, vLLM wasn't able to read the Medusa config.

abhigoyal1997 commented on vllm/worker/embedding_model_runner.py:
Changes in this file ensure that the methods in `EmbeddingModelRunner` follow the same signature as the ones in  `ModelRunner`

abhigoyal1997 commented on vllm/worker/model_runner.py:
```diff
@@ -910,11 +1000,24 @@ def capture_model(self, kv_caches: List[torch.Tensor]) -> None:
 
                 graph_runner = CUDAGraphRunner(self.model)
+
+                model_inputs = {
```
This is to generalize for models that have different input signature than transformers (like Medusa).

abhigoyal1997 commented on vllm/worker/model_runner.py:
```diff
@@ -1065,3 +1170,15 @@ def _is_block_tables_empty(block_tables: Union[None, Dict]):
     return False
+
+
+def _move_extra_tensor_data_to_seq_outputs(
```
Reorganizes the `sampled_extra_tensor_data` into sequence outputs in sampler output.

abhigoyal1997 commented on vllm/worker/model_runner.py:
```diff
@@ -630,13 +653,15 @@ def _prepare_model_input(
             num_prefill_tokens=num_prefill_tokens,
             num_decode_tokens=num_decode_tokens,
             num_prefills=num_prefills,
+            extra_inputs=extra_inputs_tensor,
```
In addition to other inputs, also prepare the extra_inputs now.

caddfa31434 commented:
@abhigoyal1997 It seems that there are some issues with TP > 1.

abhigoyal1997 commented:
@caddfa31434 Thanks for testing and catching this. The problem was with the order of execution and hidden_states broadcasting in non-driver workers. The latest commit should fix these issues (I have tested for TP <= 2).

cadedaniel commented on vllm/config.py:
```diff
@@ -321,6 +331,10 @@ def get_num_layers(self, parallel_config: "ParallelConfig") -> int:
         total_num_hidden_layers = self.hf_text_config.num_hidden_layers
         return total_num_hidden_layers // parallel_config.pipeline_parallel_size
 
+    def set_num_lookahead_tokens(self, num_lookahead_tokens: int):
```
Can you describe the relationship between num_lookahead_tokens and num medusa heads?

For context, the `num_lookahead_tokens` value is used to allocate KV space for speculative tokens. Since Medusa does not use KV, we shouldn't require this to be equal to num heads.

abhigoyal1997 replied:
This is to allow the use of lower `num_heads` than in the medusa checkpoint. In case `num_lookahead_tokens` < `num_heads`, loading the additional heads into GPU memory would be inefficient. So we set the `num_heads` = `num_lookahead_tokens` and load only the heads that are required.

cadedaniel commented on vllm/engine/arg_utils.py:
```diff
@@ -484,6 +488,13 @@ def add_cli_args(
             help=
             'The name of the draft model to be used in speculative decoding.')
 
+        parser.add_argument(
+            '--extra-inputs-for-draft-model',
+            type=nullable_str,
+            default=EngineArgs.extra_inputs_for_draft_model,
+            help='Extra model inputs used by draft model.'
+            'These should come as outputs from the target model.')
```
Can you provide an example of how this is used?

abhigoyal1997 replied:
It was used for notifying the worker that draft model takes in additional inputs e.g., for medusa, you'd need to pass "hidden_states" as the value in this argument. But, I think this can be removed. I will think of a better way which doesn't need additional command line args.

One way I am thinking is that we add an optional input_config attribute to HF configs (e.g., in `MedusaConfig`, `input_config = {"hidden_states": (hidden_states_shape, hidden_states_dtype)}`). This way, we can use just this attribute to handle the inputs and outputs of draft and target mode resp.

As you suggested, I'll create a separate PR for this which focuses only on handling the custom inputs and outputs of the models.

abhigoyal1997 replied:
This is removed now. We use model's hf_config to infer what additional inputs are required.

cadedaniel commented on vllm/engine/llm_engine.py:
```diff
@@ -245,6 +245,8 @@ def __init__(
                 ),
             ))
 
+        self.dont_speculate = False
```
What's the purpose of this state?

abhigoyal1997 replied:
This is not needed, I'll remove it.

cadedaniel commented on vllm/model_executor/models/medusa.py:
```diff
@@ -0,0 +1,110 @@
+        logits: List[torch.Tensor],
+        sampling_metadata: SamplingMetadata,
+    ) -> List[SamplerOutput]:
+        logits = torch.stack(logits, dim=0).float()
+        logprobs = torch.log_softmax(logits, dim=-1)
+        token_ids = logits.argmax(-1)  # support only top-1 for now
+        probs = torch.softmax(logits, dim=-1)
```
If we use the lossless rejection sampler, we will have to run vLLM's standard sampling routine here -- the probability distribution must be modified in the same way as the scoring probability distributions, else you will get distributional drift in the output.

abhigoyal1997 replied:
Can you please elaborate on the distribution shift? The tokens from the draft model are either accepted or rejected based on target model distribution, right? So even if the tokens from the draft are from a slightly different distribution, the final output should still match the target model distribution due to rejection. Is this understanding wrong or am I missing something?

The issue with using the standard sampling is that it was causing too much overhead. So if we do need to use it, we might need some optimizations there to get some speed-up out of Medusa.

abhigoyal1997 replied:
There's one case that I have noticed generates different tokens sometimes (not sure if this is what you are referring to though).
If without Medusa the logits of top-2 tokens have very close values (or same), then with Medusa those values sometimes change a little bit (I don't know why this is happening since Medusa shouldn't affect the output logits of the target model). This causes different tokens to be preferred by the target model, even for greedy sampling, depending on how those values change.

abhigoyal1997 replied:
I realised this was happening because of bf16 precision, not seeing any such shift when using fp32.

cadedaniel commented on vllm/sequence.py:
```diff
@@ -30,11 +30,105 @@ class Logprob:
     decoded_token: Optional[str] = None
 
 
+class ExtraTensorData:
```
In the near-future, we will separate the sequence.py datastructures from the workers, with the goal being no shared state between Worker and Scheduler unless explicitly required. The goal of this is to improve performance elsewhere in vLLM. Do we need this state to be stored in Sequence, or can it live in the Worker?

abhigoyal1997 replied:
It should be possible to put this into worker instead. Will look into this.

abhigoyal1997 replied:
This is now moved from Sequence. Any extra data is now stored inside the SpecDecodeWorker!

cadedaniel commented on vllm/sequence.py:
```diff
@@ -247,6 +349,7 @@ def __init__(
         self.read_offset = 0
         # Input + output tokens
         self.tokens: Optional[List[str]] = None
+        self.num_steps: int = 0
```
Let's integrate this with the metrics in https://github.com/vllm-project/vllm/blob/dd8de11f0a15f9ef48cd1dcac02dd8e2a8bcb494/vllm/spec_decode/metrics.py#L12

abhigoyal1997 replied:
Makes sense. Will look into this.

abhigoyal1997 replied:
Moved to metrics.

cadedaniel commented on vllm/sequence.py:
```diff
@@ -835,6 +951,12 @@ class ExecuteModelRequest:
+    # Extra inputs to pass to model
+    extra_inputs: Optional[ExtraTensorData] = None
+    # Don't speculate if true
+    dont_speculate: bool = False
```
What tests do we need this enabled for?

abhigoyal1997 replied:
Removed

cadedaniel commented on vllm/spec_decode/multi_head_worker.py:
```diff
@@ -0,0 +1,63 @@
+
+    def set_include_gpu_probs_tensor(self):
+        # Need include_gpu_probs_tensor for multi_step_worker
+        if hasattr(self.model_runner.model, "sampler"):
```
When would this be False?

abhigoyal1997 replied:
Currently `MedusaModel` doesn't have a sampler and we use the sample method itself to get the tokens. This was done because using the sampler layer was causing too much overhead.

cadedaniel commented on vllm/spec_decode/multi_head_worker.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class MultiHeadWorker(Worker):
+    """The MultiHeadWorker is equivalent to a Worker except that it allows
+    multiple forward passes in a single call, assuming the scheduler has
+    allocated enough space to store the additional KV. This reduces overhead
+    by invoking the scheduler less.
```
Do we need to inherit from the worker?

abhigoyal1997 replied:
We need to use the `init_device` method from worker. Removing inheritance from worker might require duplication of that method. Other than that, I don't think there's any dependence.

cadedaniel reviewed (commented):
Thanks for the contribution -- glad to finally see Medusa working in open-source vLLM. Adding high-level feedback. Some other questions:

* Can we add an e2e test with Medusa? we should expect greedy generation with Medusa (temp=0) to be equal to non-spec decode cases. You can follow this as an example. https://github.com/vllm-project/vllm/blob/dd8de11f0a15f9ef48cd1dcac02dd8e2a8bcb494/tests/spec_decode/e2e/test_multistep_correctness.py#L190
  * We'll also want to cover cases like cuda graph / tp>1 / other non-greedy sampling params.
* The biggest concern I have with this PR is the modification of prepare inputs for Medusa-specific models. It seems this PR is introducing two new things to prepare inputs -- allow on-GPU inputs, and also model-specific input config. Can we separate out these changes into their own PR to make things simpler? Additionally, can you walk me through the alternatives to model-specific input config in prepare inputs?

Dbxwz commented:
> Thanks for the insights into tree-mask-based version.
> As for EAGLE, I have an implementation ready for that as well. This implementation of Medusa does keep required abstractions in mind such that any speculative models which need some inputs from the target model can be implemented with ease. In fact the implementation of EAGLE I have (I'll raise a PR for that as well soon), only needs addition of a new model to vLLM without any change to the implementation logic of speculative decoding itself.

Thanks for your great work!
Based on your work, I implemented EAGLE with only  little modification. And it works well in [EAGLE-Vicuna-7B-v1.3](https://huggingface.co/yuhuili/EAGLE-Vicuna-7B-v1.3).

In the function execute_model of class ModelRunner,  the process of handling hidden_states in the prefille stage is useless for medusa and does not work for EAGLE. The prefill stage of medusa's lm_head is useless, so skipping it will be more Efficient. EAGLE‘s `FeatExtrapolator`skip the firt prompt token to keep the same shape of `inputs_embeds` and hidden_states, so I cut the hidden_states which generate the output token id and wait next spec step to get the latest hidden_states.
Synchronizing hidden_states between draft model and target model is a hard problem. To be compatible with more solutions, a more user-friendly design is necessary. Is it possible to convey additional information through extra_outputs to tell ModelRunner how to handle hidden_states ？

abhigoyal1997 commented on vllm/config.py:
```diff
@@ -137,6 +139,13 @@ def __init__(
         self.served_model_name = get_served_model_name(model,
                                                        served_model_name)
+
+        self.extra_inputs: Dict[str, Tuple[Tuple[int],
```
For medusa, this is `{"hidden_states": ((4096,), torch.bfloat16}`. The dtype can be changed (it's kept the same as dtype for the rest of the model.

abhigoyal1997 commented:
> Thanks for the contribution -- glad to finally see Medusa working in open-source vLLM. Adding high-level feedback. Some other questions:
> 
> * Can we add an e2e test with Medusa? we should expect greedy generation with Medusa (temp=0) to be equal to non-spec decode cases. You can follow this as an example. https://github.com/vllm-project/vllm/blob/dd8de11f0a15f9ef48cd1dcac02dd8e2a8bcb494/tests/spec_decode/e2e/test_multistep_correctness.py#L190
>   
>   
>   * We'll also want to cover cases like cuda graph / tp>1 / other non-greedy sampling params.
> * The biggest concern I have with this PR is the modification of prepare inputs for Medusa-specific models. It seems this PR is introducing two new things to prepare inputs -- allow on-GPU inputs, and also model-specific input config. Can we separate out these changes into their own PR to make things simpler? Additionally, can you walk me through the alternatives to model-specific input config in prepare inputs?

Thanks @cadedaniel for the review. I've replied to your comments and will work on addressing the ones that need changes. To answer your questions:

1. Yes, I'll try to add the tests.
2. For on-GPU inputs and model-specific input config I'll create a separate simplified PR. What I am thinking is that we can add an optional input_config argument in the hf_config which will hold the configuration of additional inputs. This input_config will need the name of the input, it's shape and dtype e.g., in `MedusaConfig` we add a static class attribute `input_config = {"hidden_states": (hidden_states_shape, hidden_states_dtype)}`. Let me know if this makes sense.

abhigoyal1997 commented:
> > Thanks for the insights into tree-mask-based version.
> > As for EAGLE, I have an implementation ready for that as well. This implementation of Medusa does keep required abstractions in mind such that any speculative models which need some inputs from the target model can be implemented with ease. In fact the implementation of EAGLE I have (I'll raise a PR for that as well soon), only needs addition of a new model to vLLM without any change to the implementation logic of speculative decoding itself.
> 
> Thanks for your great work! Based on your work, I implemented EAGLE with only little modification. And it works well in [EAGLE-Vicuna-7B-v1.3](https://huggingface.co/yuhuili/EAGLE-Vicuna-7B-v1.3).
> 
> In the function execute_model of class ModelRunner, the process of handling hidden_states in the prefille stage is useless for medusa and does not work for EAGLE. The prefill stage of medusa's lm_head is useless, so skipping it will be more Efficient. EAGLE‘s `FeatExtrapolator`skip the firt prompt token to keep the same shape of `inputs_embeds` and hidden_states, so I cut the hidden_states which generate the output token id and wait next spec step to get the latest hidden_states. Synchronizing hidden_states between draft model and target model is a hard problem. To be compatible with more solutions, a more user-friendly design is necessary. Is it possible to convey additional information through extra_outputs to tell ModelRunner how to handle hidden_states ？



> > Thanks for the insights into tree-mask-based version.
> > As for EAGLE, I have an implementation ready for that as well. This implementation of Medusa does keep required abstractions in mind such that any speculative models which need some inputs from the target model can be implemented with ease. In fact the implementation of EAGLE I have (I'll raise a PR for that as well soon), only needs addition of a new model to vLLM without any change to the implementation logic of speculative decoding itself.
> 
> Thanks for your great work! Based on your work, I implemented EAGLE with only little modification. And it works well in [EAGLE-Vicuna-7B-v1.3](https://huggingface.co/yuhuili/EAGLE-Vicuna-7B-v1.3).
> 
> In the function execute_model of class ModelRunner, the process of handling hidden_states in the prefille stage is useless for medusa and does not work for EAGLE. The prefill stage of medusa's lm_head is useless, so skipping it will be more Efficient. EAGLE‘s `FeatExtrapolator`skip the firt prompt token to keep the same shape of `inputs_embeds` and hidden_states, so I cut the hidden_states which generate the output token id and wait next spec step to get the latest hidden_states. Synchronizing hidden_states between draft model and target model is a hard problem. To be compatible with more solutions, a more user-friendly design is necessary. Is it possible to convey additional information through extra_outputs to tell ModelRunner how to handle hidden_states ？

For EAGLE, I was able to use the hidden_states propagation as it is without any changes. For the first draft token, the hidden_states from target model (propagated from prev. iteration) can be combined with latest sampled tokens. For the subsequent draft tokens, EAGLE itself outputs the hidden_states which can be combined with the prev sampled draft tokens. I have not added the implementation in this PR so that this is not too cluttered (though it still ended up being cluttered) but I can raise another PR for EAGLE as well once this one is final.

As for conveying additional information through extra_outputs, I don't see why that would not be possible but I am not sure exactly why and for what that's needed.

abhigoyal1997 commented on vllm/config.py:
```diff
@@ -143,6 +145,13 @@ def __init__(
         self._verify_quantization()
         self._verify_cuda_graph()
 
+        self.extra_inputs_spec: ExtraModelInputsSpec = getattr(
```
The model's config can specify if the model requires any additional inputs. This is then used in ModelRunner to capture CUDA graphs with extra inputs and execute the model.

zhyncs reviewed (commented):
It is recommended to include the statistics of the accept rate as well.

zhyncs commented:
> It is recommended to include the statistics of the accept rate as well.

Assuming the medusa head num is 5, during each iteration, for example, if out of the 5 results generated in the first round, 3 were accepted and in the second round, 2 were accepted, there would be two data points: 60% and 40%. When finally displaying statistics, it should be a curve showing how the accept rate changes with each iteration after using medusa for a particular sequence.

zhyncs commented:
You may also add the test results of Llama3 8B Instruct https://huggingface.co/leptonai/Llama-3-8b-instruct-4-heads, trained by Lepton AI. They use it in production environments. Vicuna 13B is a bit outdated.

cadedaniel commented:
I will take another pass on Monday (out of office this week).

zhyncs commented:
> what adjustments specifically need to be done on the Cos/Sin matrix for RoPE?

For a normal input, such as `a,b,c,d,e` the first two columns of the corresponding cos matrix are as follows

token | first | second
-- | -- | --
a | cos(0*θ0) | cos(0*θ1)
b | cos(1*θ0) | cos(1*θ1)
c | cos(2*θ0) | cos(2*θ1)
d | cos(3*θ0) | cos(3*θ1)
e | cos(4*θ0) | cos(4*θ1)

For a medusa input, such as `r, a, b, c, d, e, c, d, e`

token | first | second
-- | -- | --
r | cos(0*θ0) | cos(0*θ1)
a | cos(1*θ0) | cos(1*θ1)
b | cos(1*θ0) | cos(1*θ1)
c | cos(2*θ0) | cos(2*θ1)
d | cos(2*θ0) | cos(2*θ1)
e | cos(2*θ0) | cos(2*θ1)
c | cos(2*θ0) | cos(2*θ1)
d | cos(2*θ0) | cos(2*θ1)
e | cos(2*θ0) | cos(2*θ1)

`r` belongs to `root`, `a` and `b` belong to `head 0`, `c` `d` and `e` belong to `head 1`.

Due to the inconsistency between the position of tokens in the sequence dimension and the semantic logic dimension, Medusa's cos/sin needs to be adjusted according to the semantic logic dimension.

This part is the cos matrix for the current `seq_q`, while the cos matrix for `cache_k` has tokens arranged in a determined order.

abhigoyal1997 commented:
@cadedaniel I've refactored how the extra outputs/inputs are handled. `TensorData` is now stored in `SpecDecodeWorker` instead of `SequenceData` and the preparation is also decoupled from `_prepare_model_input` (it's done separately in `_prepare_extra_model_input`). I've also removed the CLI changes and now the information about extra inputs required by the draft model is fetched from the `ModelConfig.hf_config` e.g., for Medusa, it lives in `MedusaConfig`.

I've also added these input preparation changes as a separate PR #5101.

cadedaniel commented:
Going to give another pass tomorrow.

abhigoyal1997 commented:
Hi @LiuXiaoxuanPKU 
It's mostly ready but since there have been some additions to the main branch, it might need a refactor to incorporate those. As for the results, the first 2 rows are numbers without Medusa (or any speculative decoding) and next 2 rows are with Medusa. Same for Meta llama version.
I'll refactor soon and resolve the conflicts with main. Will let you know once this is done.

cadedaniel commented:
@abhigoyal1997 and @njhill -- seems there's some overlap in how your PRs share state between forward passes. What do we think is a good way forward?

njhill commented:
@cadedaniel we were going for a minimally-invasive and less general approach in the #4947. IMHO this would be preferred as a first step until we have more concrete examples of where arbitrary inputs/outputs are needed (which might then further inform the abstraction design).

@abhigoyal1997 WDYT about a simpler version of your PR that makes use of the hidden states handling that we have in #4947?

sighingnow commented:
> Hi @caddfa31434 Tree-style speculation and verification are certainly something I am interested in. But for complete benefit, I was still waiting on the support for tree attention. I have implemented it in a PyTorch implementation but haven't thought about how it would work in vLLM.

FYI: flashinfer supports custom attention mask in https://github.com/flashinfer-ai/flashinfer/pull/266.

abhigoyal1997 commented:
> @cadedaniel we were going for a minimally-invasive and less general approach in the #4947. IMHO this would be preferred as a first step until we have more concrete examples of where arbitrary inputs/outputs are needed (which might then further inform the abstraction design).
> 
> @abhigoyal1997 WDYT about a simpler version of your PR that makes use of the hidden states handling that we have in #4947?

@njhill Makes sense. I kept it more general as I was trying out various approaches for draft models that use different inputs. But I agree, we haven't seen anything other than the usual LLM inputs + hidden_states being used in draft models. I can refactor this PR to keep things minimal and then we can add on as required later on.

abhigoyal1997 committed:
Minimal implementation of Medusa

abhigoyal1997 committed:
formatting

abhigoyal1997 commented:
@cadedaniel @LiuXiaoxuanPKU @njhill This is ready for a review. I have refactored so that this PR only adds minimal changes required to run Medusa draft models.

njhill commented on vllm/worker/model_runner.py:
```diff
@@ -774,9 +774,8 @@ def execute_model(
-            if seq_group_metadata_list[0].is_prompt:
-                hidden_states = hidden_states.index_select(
-                    0, sampling_metadata.selected_token_indices)
+            hidden_states = hidden_states.index_select(
```
I don't think it would break anything just might be a slight extra overhead for eager case

abhigoyal1997 replied:
In that case, we can replace this with just simple slicing in case of graph execution + decode.

njhill replied:
@abhigoyal1997 that sounds good!

abhigoyal1997 committed:
Merge branch 'vllm-project:main' into medusa

njhill commented on vllm/config.py:
```diff
@@ -916,6 +916,10 @@ def maybe_create_spec_config(
                     "Speculative decoding with mlp_speculator models does not "
                     "yet support distributed inferencing (TP > 1).")
 
+            if draft_model_config.hf_config.model_type == "medusa":
```
I realized that I had not properly wired this up in the MLPSpeculator impl, will make a fix for that!

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into medusa

abhigoyal1997 committed:
fixes after merge

abhigoyal1997 commented on vllm/worker/model_runner.py:
```diff
@@ -1024,6 +1024,11 @@ def execute_model(
                 assert model_input.sampling_metadata is not None
                 hidden_states = hidden_states.index_select(
                     0, model_input.sampling_metadata.selected_token_indices)
+            elif decode_meta.use_cuda_graph:
```
For graph execution in decode step, this removes padding in hidden states.

abhigoyal1997 committed:
Merge branch 'main' into medusa

njhill commented on vllm/config.py:
```diff
@@ -927,6 +927,10 @@ def maybe_create_spec_config(
 
+            if draft_model_config.hf_config.model_type == "medusa":
+                draft_model_config.hf_config.set_num_lookahead_tokens(
+                    num_speculative_tokens)
```
@abhigoyal1997 I just merged https://github.com/vllm-project/vllm/pull/5876 which fixes the equivalent thing for MLPSpeculator.

Was thinking you could adjust the medusa config to expose `num_lookahead_tokens` as a attribute/property instead of a custom setter (could use python property annotation).. then hopefully the code to set it here can be common and doesn't need to check the explicit model type. WDYT?

abhigoyal1997 replied:
modified.

abhigoyal1997 committed:
minor fix

abhigoyal1997 committed:
Merge branch 'medusa' of github.fkinternal.com:abhinav-goyal/vllm into medusa

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into medusa

abhigoyal1997 committed:
fix for medusa head loading

abhigoyal1997 committed:
Merge branch 'medusa' of github.fkinternal.com:abhinav-goyal/vllm into medusa

abhigoyal1997 committed:
formatting

abhigoyal1997 committed:
minor fix for medusa heads = 1

abhigoyal1997 committed:
Merge branch 'main' of github.fkinternal.com:abhinav-goyal/vllm into medusa

abhigoyal1997 committed:
minor fixes after merge

abhigoyal1997 commented:
@LiuXiaoxuanPKU Can you please take a look?

LiuXiaoxuanPKU reviewed (changes requested):
Sorry for the late review! LGTM! But I have some minor questions & comments:
1. It seems only greedy sampling is supported. Standard sampling is not supported.
2. Could you add some end to end tests to make sure medusa generates almost same results as without speculative decoding? No need to be very strict. Take a look at [this](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_mlp_correctness.py).

Happy to get this pr merged soon, sorry for the delay!

abhigoyal1997 commented:
Thanks @LiuXiaoxuanPKU for the review!

> Sorry for the late review! LGTM! But I have some minor questions & comments:
> 
> 1. It seems only greedy sampling is supported. Standard sampling is not supported.

Standard sampling works as well, I have the results showing speed-up with temperature=1 in the PR description. Since I have only added the top-1 proposal candidate in this PR, sampling in the Medusa model is just using the argmax (we can extend this when we add tree speculation).  My thought was that even when sampling with a temperature of 1, the target model would more likely choose the top-1 token because that still has the highest probability, so if we choose the top-1 token from Medusa, it still has more chance of being accepted than other tokens (the chance that token sampled from target and Medusa head would match when both are random would be much lower). Am I missing something in this argument? Even in Medusa paper, the candidates are only formed by top-k tokens.

> 2. Could you add some end to end tests to make sure medusa generates almost same results as without speculative decoding? No need to be very strict. Take a look at [this](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_mlp_correctness.py).

Yes, I'll add the tests. Thanks for the reference!

> 
> Happy to get this pr merged soon, sorry for the delay!

abhigoyal1997 committed:
Adding E2E tests for Medusa

abhigoyal1997 committed:
Changing base model to a smaller model in test_medusa_correctness to prevent OOM in CI

abhigoyal1997 committed:
Replacing Medusa models in test with smaller ones to prevent OOM

hustxiayang commented:
Hi, I assume that you got Mistral-7B-Instruct-v0.2 + medusa from https://huggingface.co/text-generation-inference/Mistral-7B-Instruct-v0.2-medusa, and for Meta-Llama-3-8B-Instruct + medusa, you trained Mudusa-1 by yourself using the codebases from https://github.com/FasterDecoding/Medusa/tree/main?

hustxiayang commented:
Hi, it seems that compared to the speedups of MLPspeculator, the speedups are much smaller: https://github.com/vllm-project/vllm/pull/4947. Do you have any insights on when we should use medusa instead of MLPspeculator?

abhigoyal1997 commented:
> Hi, I assume that you got Mistral-7B-Instruct-v0.2 + medusa from https://huggingface.co/text-generation-inference/Mistral-7B-Instruct-v0.2-medusa, and for Meta-Llama-3-8B-Instruct + medusa, you trained Mudusa-1 by yourself using the codebases from https://github.com/FasterDecoding/Medusa/tree/main, but how many medusa heads you used for training?

Hi @hustxiayang 
Actually I trained my own Medusa heads for both Mistral and Llama3 (will soon release them on HF as well). As for the number of heads, I used 4.

> Hi, it seems that compared to the speedups of MLPspeculator, the speedups are much smaller: #4947. Do you have any insights on when we should use medusa instead of MLPspeculator?

I am not sure whether there are specific guidelines on when one should be used over the other. As for the performance, it depends on the dataset used, training method etc. and the results I posted in the description are for a baseline model I trained which can be improved with more work. Tree speculation and verification using MQA should improve the performance as well.

hustxiayang commented:
Hi, it seems that this implementation is based on Medusa version 1, which loads lm_heads for each medusa heads. In Medusa version 2, it proposed to reuse the lm_heads from the base model. Have you already investigated on this and do you plan to implement it? Thanks!
